### PR TITLE
feat: basic validation of CAR files

### PR DESCRIPTION
**Pull request by lemmih**
## Summary of changes

Changes introduced in this pull request:

- add `forest-tool car validate` command.

This new command verifies the invariants expected to be true for all CAR files (not just Filecoin CAR files).

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/tool/subcommands/car_cmd.rs [hunk 1]
```diff
@@ -50,7 +69,51 @@ impl CarCommands {
                 crate::db::car::forest::Encoder::write(&mut writer, all_roots, frames).await?;
                 writer.flush().await?;
             }
+            Self::Validate {
+                car_file,
+                ignore_block_validity: check_block_validity,
```
Comment (hanabi1224) [inline src/tool/subcommands/car_cmd.rs:74]: Should `check_block_validity` be `!ignore_block_validity` here?

File: src/tool/subcommands/car_cmd.rs [hunk 2]
```diff
@@ -50,7 +69,51 @@ impl CarCommands {
...
+
+// At present, three invariants are checked:
+// - The CAR file is syntactically valid and all blocks can be streamed.
+// - Each block CID is checked against the hash of the block.
+// - Each block CID is looked-up in the on-disk index.
+//
+// Invariants related to Filecoin are not checked. For those, see `forest-tool
+// snapshot validate`.
+//
+// We do not check for duplicate blocks. Whether duplicate blocks are allowed or
```
Comment (LesnyRumcajs) [inline src/tool/subcommands/car_cmd.rs:90]: Should we have some tests for this feature? Or in compiler and infallible code we trust?

Comment (lemmih) [reply src/tool/subcommands/car_cmd.rs]: Done. All the error cases are now tested.

File: src/tool/subcommands/car_cmd.rs [hunk 3]
```diff
@@ -50,7 +69,166 @@ impl CarCommands {
...
+    // If a CarBlock exist that isn't referenced in the index, this is an error.
+    #[tokio::test]
+    async fn validate_invalid_index() {
+        let data = "this data _does_ match the CID".as_bytes().to_vec();
+        let cid = Cid::new_v1(0, Code::Blake2b256.digest(&data));
+        let temp_path = create_raw_car_file(vec![CarBlock { cid, data }], vec![cid]).await;
+
+        assert!(validate(&temp_path, false, false).await.is_err());
+    }
+}
```
Comment (LesnyRumcajs) [inline src/tool/subcommands/car_cmd.rs:234]: Short, concise tests. That's just code porn.

File: src/tool/subcommands/car_cmd.rs [hunk 4]
```diff
@@ -50,7 +69,166 @@ impl CarCommands {
...
+    use cid::multihash::{Code, MultihashDigest};
+    use cid::Cid;
+    use futures::{stream::iter, StreamExt, TryStreamExt};
+    use std::io::Write;
+    use tempfile::{NamedTempFile, TempPath};
+    use tokio::io::AsyncWriteExt;
+
+    #[tokio::test]
+    async fn validate_junk_car() {
+        let mut temp_path = NamedTempFile::new_in(".").unwrap();
```
Comment (LesnyRumcajs) [inline src/tool/subcommands/car_cmd.rs:136]: Why does it need to be in `.`? In general, I'd prefer that we don't create test artefacts in the source code directory, even temporarily.

**Review by hanabi1224 — APPROVED** [review]

File: src/tool/subcommands/car_cmd.rs (hunk 4 above)
Comment (lemmih) [reply src/tool/subcommands/car_cmd.rs:136]: It doesn't. But you banned `NamedTempFile::new`.

Comment (LesnyRumcajs) [reply src/tool/subcommands/car_cmd.rs:136]: How about `tempfile::Builder::new().tempfile().unwrap()`?

Comment (lemmih) [reply src/tool/subcommands/car_cmd.rs]: Done.

**Review by LesnyRumcajs — APPROVED** [review]
We could also cover the case where we ignore the block validity and the Forest index. All in all, a user could do this. (not sure why, though)

**Comment by lemmih** [comment]
> We could also cover the case where we ignore the block validity and the Forest index. All in all, a user could do this. (not sure why, though)

Done.

**Review by LesnyRumcajs — APPROVED** [review]

**Pull request merged**